File: python3/1233.py
from typing import List

class Solution:

    def subfolder(self, folder: str, dirs: set) -> bool:

        i = 0
        while True:
            i = folder.find('/', i + 1)
            if i == -1:
                return False

            if folder[0: i] in dirs:
                return True

    def removeSubfolders(self, folder: List[str]) -> List[str]:

        dirs = set(folder)

        for f in folder:
            if self.subfolder(f, dirs):
                dirs.remove(f)

        return list(dirs)

File: rust/1233.rs
// 1233. Remove Sub-Folders from the Filesystem

use std::collections::HashSet;

impl Solution {
    pub fn is_subfolder(folder: &str, dirs: &HashSet<&str>) -> bool {
        let mut i = 0;

        loop {
            i = match folder[i + 1..].find('/') {
                Some(index) => index + i + 1,
                None => return false,
            };

            if dirs.contains(&folder[0..i]) {
                return true;
            }
        }
    }

    pub fn remove_subfolders(folder: Vec<String>) -> Vec<String> {
        let mut dirs: HashSet<&str> = folder.iter().map(|f| f.as_str()).collect();

        for f in &folder {
            if Self::is_subfolder(f, &dirs) {
                dirs.remove(f.as_str());
            }
        }

        dirs.into_iter().map(|f| f.to_string()).collect()
    }
}
